### PR TITLE
feat(api): Create all mewm/event type endpoints

### LINKED ISSUE #73: [MEWM] Develop, Test, and Document MEWM Event Type API
**Describe the solution you'd like**
An API should be created to handle basic MEWM Event Type operations. This API should include the following endpoints:

- GET mewm/types/ - @shriyash 
- GET mewm/types/[typeId] - @shriyash 
- POST mewm/types/ - @shriyash 
- PATCH mewm/types/[typeId] - @shriyash 
- DELETE  mewm/typet/[typeId] - @shriyash 

**Additional context**
_MEWM Goal:_

The goal of this module would be to provide an all-inclusive center of information for the mini-events and workshops occurring at the hackathon. This module could be used by all users that are associated with the event, albeit a hacker, organizer, sponsor, etc. Different types of users will have different permissions and functionalities in the module but the goal is relatively the same for all of them, which is to help facilitate user experience with the events and workshops. Note: all events should be displayed in the user’s local time once the time slot is created!

_Key Features in this Item:_

- Organizers can create event visibility levels and categories (ex. Internal only, meal, deadline, workshop, mini-event, etc.) and assign who can view them by permission role

### PR DESCRIPTION
**Pull request by shriyash**
Closes #73

### CODE REVIEW

**Comment by nicksinishtaj** [comment]
Can you quickly do the following:

Copy paste this comment I have right below here, and then update with your endpoints and the data you're using to verify in the API explorer. Just makes it easier when referencing back to these what test data to run.

Example is from my PR and what I used to test the endpoints

### Tests via API Explorer (ORGANIZER user):
**POST mewm/events/**

```JSON
{
    "eventId": "event1",
    "eventName": "Ejub's Excel Contest",
    "eventDescription": "Ejub displays his magnificent skill of beautiful art in Microsoft Excel. They refer to Ejub as the Picasso of Spreadsheets ",
    "virtual": true
}
```

> Status: 201

**GET mewm/events/event1**

mewm/events/event1

> Status: 200
> Body: {"eventId":"event1","virtual":true,"eventDescription":"Ejub displays his magnificent skill of beautiful art in Microsoft Excel. They refer to Ejub as the Picasso of Spreadsheets ","eventName":"Ejub's Excel Contest"}

**PATCH mewm/events/event1**

```JSON
{
	"eventName": "ejub excel"
}
```

> Status: 200

**GET mewm/events/event1**

> Status: 200
> Body: {"eventId":"event1","virtual":true,"eventDescription":"Ejub displays his magnificent skill of beautiful art in Microsoft Excel. They refer to Ejub as the Picasso of Spreadsheets ","eventName":"ejub excel"}

**POST mewm/events/**

```JSON
{
    "eventId": "event2",
    "eventName": "Ejub's LIVE Excel Contest",
    "eventDescription": "Ejub displays his magnificent skill of beautiful art in Microsoft Excel, but in real-time on a white board. They refer to Ejub as the Picasso of Spreadsheets ",
    "virtual": false
}
```

> Status: 201

**GET mewm/events/**

> Status: 200
> Body: {"miniEvents":[{"eventId":"event2","virtual":false,"eventDescription":"Ejub displays his magnificent skill of beautiful art in Microsoft Excel, but in real-time on a white board. They refer to Ejub as the Picasso of Spreadsheets ","eventName":"Ejub's LIVE Excel Contest"},{"eventId":"event1","virtual":true,"eventDescription":"Ejub displays his magnificent skill of beautiful art in Microsoft Excel. They refer to Ejub as the Picasso of Spreadsheets ","eventName":"ejub excel"}]}



**DELETE mewm/events/event1**

> Status: 200

**GET mewm/events/**

> Status: 200
> Body: {"miniEvents":[{"eventId":"event2","virtual":false,"eventDescription":"Ejub displays his magnificent skill of beautiful art in Microsoft Excel, but in real-time on a white board. They refer to Ejub as the Picasso of Spreadsheets ","eventName":"Ejub's LIVE Excel Contest"}]}

**Comment by shriyash** [comment]
Tests via API Explorer (ORGANIZER user):
### POST mewm/types/
```
{
    "eventTypeId": "eventType1",
    "eventTypeName": "Workshop",
    "eventTypeDescription": "A technical talk where a speaker will guide attendees through an introduction or exploration of a topic"
},
{
    "eventTypeId": "eventType2",
    "eventTypeName": "Fun",
    "eventTypeDescription": "A fun event where you can relax and meet others."
}
```

> Status: 201

### GET mewm/types/

Result:
```
{
   "eventTypes":[
      {
         "eventTypeDescription":"A fun event where you can relax and meet others.",
         "eventTypeId":"eventType2",
         "eventTypeName":"Fun"
      },
      {
         "eventTypeId":"eventType1",
         "eventTypeName":"Workshop",
         "eventTypeDescription":"A technical talk"
      }
   ]
}
```

> Status: 200

### GET mewm/types/eventType1

Result:
```
{
   "eventTypeId":"eventType1",
   "eventTypeName":"Workshop",
   "eventTypeDescription":"A technical talk"
}
```

> Status: 200


### PATCH mewm/types/eventType1
```
{
	"eventTypeDescription": "A technical talk"
}
```
> Status: 200

### DELETE mewm/types/eventType2

> Status: 200

File: @types/modules/mewm/types.d.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+declare interface MEWMEventTypeUpdateRequest {
+  eventTypeName?: string;
+  eventTypeDescription?: string;
```
Comment (andrewdimmer) [inline @types/modules/mewm/types.d.ts:3]: Can we add the `permissions` object here?
```suggestion
  eventTypeDescription?: string;
  permissions?: UASPermissionSwitchConfig<boolean>;
```

File: @types/modules/mewm/types.d.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+declare interface MEWMEventTypeCreateRequest
+  extends MEWMEventTypeUpdateRequest {
+  eventTypeId: string;
+  eventTypeName: string;
+}
+
+declare interface MEWMEventType extends MEWMEventTypeCreateRequest {
+  eventTypeDescription: string;
```
Comment (andrewdimmer) [inline @types/modules/mewm/types.d.ts:13]: Can we add the `permissions` object here?
```suggestion
  eventTypeDescription: string;
  permissions: UASPermissionSwitchConfig<boolean>;
```

File: functions/src/modules/mewm/types/createType.ts [hunk 3]
```diff
@@ -1,5 +1,44 @@
...
+};
 
-export default replaceMe;
+const validate: ExpressFunction = (req, res, next) => {
+  const validationRules: ValidatorObjectRules = {
+    type: "object",
+    rules: {
+      eventTypeId: { rules: ["string"], required: true },
+      eventTypeName: { rules: ["string"], required: true },
+      eventTypeDescription: { rules: ["string", "emptystring"] },
```
Comment (andrewdimmer) [inline functions/src/modules/mewm/types/createType.ts:21]: Can we add the permission object here?

File: functions/src/modules/mewm/types/createType.ts [hunk 4]
```diff
@@ -1,5 +1,44 @@
...
+    },
+  };
+  requestBodyTypeValidator(req, res, next)(validationRules, execute);
+};
+
+const execute: ExpressFunction = (req, res, next) => {
+  const errorHandler = expressErrorHandlerFactory(req, res, next);
+
+  const body = res.locals.parsedBody as MEWMEventTypeCreateRequest;
+
```
Comment (andrewdimmer) [inline functions/src/modules/mewm/types/createType.ts:31]: Can we autofill the description to an empty string and the permissions to an empty object if none is provided?

File: functions/src/modules/mewm/types/listTypes.ts [hunk 5]
```diff
@@ -1,5 +1,31 @@
...
+      const eventTypes: MEWMEventType[] = [];
+      for (const doc of documents.docs) {
+        eventTypes.push(doc.data() as MEWMEventType);
+      }
+      res
+        .status(200)
+        .send(JSON.stringify({ eventTypes } as MEWMEventTypeList));
+      next();
+    })
+
```
Comment (andrewdimmer) [inline functions/src/modules/mewm/types/listTypes.ts:27]: Minor nitpick, but can we remove this extra line as long as we're making other changes?
```suggestion
```

File: functions/src/modules/mewm/types/updateType.ts [hunk 6]
```diff
@@ -1,5 +1,42 @@
...
+  })(req, res, next);
+};
 
-export default replaceMe;
+const validateOrganizer: ExpressFunction = (req, res, next) => {
+  const validationRules: ValidatorObjectRules = {
+    type: "object",
+    rules: {
+      eventTypeName: { rules: ["string"] },
+      eventTypeDescription: { rules: ["string", "emptystring"] },
```
Comment (andrewdimmer) [inline functions/src/modules/mewm/types/updateType.ts:20]: Can we add the permissions object here?

**Review by andrewdimmer — CHANGES_REQUESTED** [review]
Looks really good, but I think we're missing the `permissions` object from the data type, create, and update functions. Can you take a look at that for me?

**Review by andrewdimmer — APPROVED** [review]
Looks good to me!

**Pull request merged**